stdpopsim/models.py

```diff
@@ -137,6 +137,176 @@ def asdict(self):
+    def get_Ne_through_time_single_pop(self, end, num_steps=10):
+        """
+        This function computes the defined Ne for each individual
+        sub population for any demographic model.
+
+        `end` specifies at which generation to stop
+        collecting effective population sizes among subpopulations
+
+        `num_steps` parameter will determine how many points between
+        `start` and `end` to sample.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the effective population sizes for
+        each subpopulation individually, for each step, as defined
+        by the model which inherits from this class.
+        """
+
+        num_pops = self.num_pops()
+        N_t = np.zeros([num_steps, num_pops])
+        steps = np.linspace(0, end, num_steps)
+        for j, t in enumerate(steps):
+            N = self._get_N_M_at_t(t)["population_sizes"]
+            N_t[j] = N
+        return steps, N_t
+
+    # TODO Rename
+    def get_Ne_through_time_all_pops(self, end, num_samples_per_location, num_steps=10):
+        """
+        This function will calculate the ground truth
+        effective population size (Ne) for a population with
+        multiple sub-populations. This takes a
+        demographic model and returns a
+        function of time that gives inverse coalescence rate.
+
+        `num_samples_per_location` should be a list containing the
+        number of samples in each subpopulation organized by population index.
+
+        `num_steps` parameter will determine how many points along the
+        time axis are returnes.
+
+        Returns a tuple where the first value
+        is the steps (in generations) and the second is a numpy
+        ndarray that will contain the effective population sizes for
+        each subpopulation individually, for each step, as defined
+        by the model which inherits from this class.
+        """
+
+        num_pops = self.num_pops()
+        assert(len(num_samples_per_location) == num_pops)
+        P = np.zeros([num_pops**2, num_pops**2])
+        index_array = np.array(range(num_pops**2)).reshape([num_pops, num_pops])
+        for x in range(num_pops):
+            for y in range(num_pops):
+                K_x = num_samples_per_location[x]
+                K_y = num_samples_per_location[y]
+                P[index_array[x, y], index_array[x, y]] = K_x * (K_y - self._delta(x, y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        for j in range(num_steps):
+            N_M = self._get_N_M_at_t(steps[j])
+            pop_sizes = N_M["population_sizes"]
+            mig_matrix = N_M["migration_matrix"]
+            C = self._get_C_from_N(pop_sizes)
+            Mp = self._get_Mp_from_M(mig_matrix)
+            G = self._get_G_from_Mp_C(Mp, C)
+            P = P * np.exp(dt * G)
```

Comment:
This needs to be (a) matrix multiplication, and (b) matrix exponentiation, instead of the "usual" ones. So:
```
P = np.matmul(P, scipy.linalg.expm(dt * G))
```

Reply:
Ahhh, That might mess up the multi-loc for sure.